java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -84,6 +85,36 @@ public JavaType resolveTypeSubstitution(JavaType type, JavaType definition) {
     return typeSubstitutionSolver.applySiteSubstitution(type, definition);
   }
 
+  public JavaType resolveTypeSubstitutionWithDiamondOperator(ParametrizedTypeJavaType type, JavaType definition) {
+    if (isParametrizedType(definition)) {
+      ParametrizedTypeJavaType parametrizedDefinition = (ParametrizedTypeJavaType) definition;
+      JavaType newType = type;
+      if (type.rawType == parametrizedDefinition.rawType) {
+        newType = typeSubstitutionSolver.applyOnParametrizedTypeWithoutSubsitution(type, parametrizedDefinition.typeSubstitution);
+      } else if (type.symbol.typeVariableTypes.size() == parametrizedDefinition.typeSubstitution.size()){
+        // copy the substitution from the definition type to get a substitution using adequate type variables
+        TypeSubstitution substitution = new TypeSubstitution();
+        List<JavaType> substitutedTypes = parametrizedDefinition.typeSubstitution.substitutedTypes();
+        for (int i = 0; i < type.symbol.typeVariableTypes.size(); i++) {
+          TypeVariableJavaType var = type.symbol.typeVariableTypes.get(i);
+          substitution.add(var, substitutedTypes.get(i));
+        }
+        newType = typeSubstitutionSolver.applyOnParametrizedTypeWithoutSubsitution(type, substitution);
+      } else {
+        // FIXME SONARJAVA-? should be deduced from the supertypes of type, based on substitution provided by definition
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1134)